### PR TITLE
getlantern/lantern#2274 getlantern/lantern#2276 Cloud configuration fixe...

### PR DESCRIPTION
**Pull request by oxtoacart**
...s

### CODE REVIEW

File: src/github.com/getlantern/flashlight/config/config.go [hunk 1]
```diff
@@ -204,6 +206,10 @@ func (cfg *Config) ApplyDefaults() {
 		cfg.UIAddr = "localhost:16823"
 	}
 
+	if cfg.CloudConfig == "" {
+		cfg.CloudConfig = "https://s3.amazonaws.com/lantern_config/cloud.valencia.yaml.gz"
+	}
+
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/config.go:212]: I realized that we weren't defaulting the cloud config URL anywhere, so I added that.

File: src/github.com/getlantern/flashlight/config/config.go [hunk 2]
```diff
@@ -332,10 +338,20 @@ func (cfg Config) fetchCloudConfig() (bytes []byte, err error) {
...
 func (cfg Config) doFetchCloudConfig(proxyAddr string) ([]byte, error) {
 	log.Tracef("doFetchCloudConfig via '%s'", proxyAddr)
+
+	if proxyAddr != "" {
+		// Wait for proxy to become available
+		err := waitforserver.WaitForServer("tcp", proxyAddr, 30*time.Second)
+		if err != nil {
+			return nil, fmt.Errorf("Proxy never came up at %v: %v", proxyAddr, err)
+		}
+	}
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/config.go:348]: This takes care of getlantern/lantern#2274 by waiting for the proxy before trying to download via it.

File: src/github.com/getlantern/flashlight/config/config.go [hunk 3]
```diff
@@ -393,6 +411,7 @@ func (updated *Config) updateFrom(updateBytes []byte) error {
 		for domain, _ := range wlDomains {
 			updated.ProxiedSites.Cloud = append(updated.ProxiedSites.Cloud, domain)
 		}
+		sort.Strings(updated.ProxiedSites.Cloud)
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/config.go:414]: For readability and better diffs, I'm sorting the ProxiedSites.cloud list everywhere now.

File: src/github.com/getlantern/flashlight/genconfig/cloud.yaml.tmpl [hunk 4]
```diff
@@ -16,13 +16,13 @@ client:
...
     qos: 2
     weight: 80
-  proxiedsites:
-    cloud:{{range $domain, $bool := .proxiedsites }}
-    - {{$domain}}{{end}}
   masqueradesets:
     cloudflare:{{range .masquerades}}
     - domain: {{.Domain}}
       ipaddress: {{.IpAddress}}{{end}}
+proxiedsites:
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/genconfig/cloud.yaml.tmpl:23]: This was the fix for getlantern/lantern#2276.  I had moved the proxiedsites configuration out of the client section, but forgot to update the template.  I've uploaded the fixed yaml file to Amazon already.

File: src/github.com/getlantern/flashlight/genconfig/genconfig.go [hunk 5]
```diff
@@ -265,10 +265,15 @@ func buildModel(cas map[string]*castat, masquerades []*masquerade) map[string]in
 	}
 	sort.Sort(ByFreq(casList))
 	sort.Sort(ByDomain(masquerades))
+	ps := make([]string, 0, len(proxiedSites))
+	for site, _ := range proxiedSites {
+		ps = append(ps, site)
+	}
+	sort.Strings(ps)
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/genconfig/genconfig.go:272]: Sorting proxied sites before generating yaml and built-in config.

**Comment by atavism** [comment]
LGTM!

**Pull request merged**